client/beefy/src/worker.rs

```diff
@@ -119,209 +134,190 @@ where
-				metric_set!(self, beefy_best_block, *notification.header.number());
-			}
+					.notify(|| Ok::<_, ()>(hash))
+					.expect("forwards closure result; the closure always returns Ok; qed.")
+			});
+			// Set new best BEEFY block number.
+			self.best_beefy_block = Some(block_num);
+			metric_set!(self, beefy_best_block, block_num);
+		} else {
+			debug!(target: "beefy", "🥩 Can't set best beefy to older: {}", block_num);
 		}
+	}
 
-		if self.should_vote_on(*notification.header.number()) {
-			let (validators, validator_set_id) = if let Some(rounds) = &self.rounds {
-				(rounds.validators(), rounds.validator_set_id())
-			} else {
-				debug!(target: "beefy", "🥩 Missing validator set - can't vote for: {:?}", notification.header.hash());
-				return
-			};
-			let authority_id = if let Some(id) = self.key_store.authority_id(validators) {
-				debug!(target: "beefy", "🥩 Local authority id: {:?}", id);
-				id
-			} else {
-				debug!(target: "beefy", "🥩 Missing validator id - can't vote for: {:?}", notification.header.hash());
-				return
-			};
-
-			let mmr_root =
-				if let Some(hash) = find_mmr_root_digest::<B, Public>(&notification.header) {
-					hash
-				} else {
-					warn!(target: "beefy", "🥩 No MMR root digest found for: {:?}", notification.header.hash());
-					return
-				};
-
-			let payload = Payload::new(known_payload_ids::MMR_ROOT_ID, mmr_root.encode());
-			let commitment = Commitment {
-				payload,
-				block_number: notification.header.number(),
-				validator_set_id,
-			};
-			let encoded_commitment = commitment.encode();
+	/// Handle session changes by starting new voting round for mandatory blocks.
+	fn init_session_at(&mut self, active: ValidatorSet<AuthorityId>, session_start: NumberFor<B>) {
+		debug!(target: "beefy", "🥩 New active validator set: {:?}", active);
+		metric_set!(self, beefy_validator_set_id, active.id());
+		// BEEFY should produce a signed commitment for each session
+		if active.id() != self.last_signed_id + 1 && active.id() != GENESIS_AUTHORITY_SET_ID {
+			metric_inc!(self, beefy_skipped_sessions);
+		}
 
-			let signature = match self.key_store.sign(&authority_id, &*encoded_commitment) {
-				Ok(sig) => sig,
-				Err(err) => {
-					warn!(target: "beefy", "🥩 Error signing commitment: {:?}", err);
-					return
-				},
-			};
+		if log_enabled!(target: "beefy", log::Level::Debug) {
+			// verify the new validator set - only do it if we're also logging the warning
+			let _ = self.verify_validator_set(&session_start, &active);
+		}
 
-			trace!(
-				target: "beefy",
-				"🥩 Produced signature using {:?}, is_valid: {:?}",
-				authority_id,
-				BeefyKeystore::verify(&authority_id, &signature, &*encoded_commitment)
-			);
+		let prev_validator_set = if let Some(r) = &self.rounds {
+			r.validator_set().clone()
+		} else {
+			// no previous rounds present use new validator set instead (genesis case)
+			active.clone()
```

Comment:
Please check my thoughts:
1) we have some chain that has been started a year ago, so we're definitely not at genesis;
2) authority node is starting and `BeefyWorker::new()` is called, which sets `rounds` to `None`;
3) worker receives finality notification - i.e. `handle_finality_notification` is called;
4) the finalized header starts new session - i.e. `handle_finality_notification` calls `init_session_at`;
5) since `rounds` is still set to `None`, we're falling here and prev validator set is set to the same (new) set.

I don't know if I'm right and if I am, then whether it may lead to some issues or not. So just asking to recheck it. Thanks :)

Reply:
Your reasoning is correct, but we plan on addressing the `initial sync` part after Lean Beefy milestone - issue for it here https://github.com/paritytech/grandpa-bridge-gadget/issues/112.

Within Lean Beefy simplifying assumptions, [this happens](https://github.com/acatangiu/substrate/blob/d807c65ce89f01ccc8975f114a4fb4b526f3a172/client/beefy/src/worker.rs#L477) when starting not at genesis.

The end result is that this validator will not be able to vote for current session (doesn't have the right validator set), but will be able to vote for future ones.